File: Grokking Algorithms in Python/Chapter 3 .py
#RECURSION
#function to print numbers from N to 1 using recursion
def countDown(n):
    if(n <= 0):             #base case:
        return
    else:
        print(n)
        countDown(n-1)      #recursive case:

#function to find the factorial of the number using recursion
def factorial(n):
    if(n == 1 or n == 0):
        return 1            #base case (the point which the program will stop when it meets it)
    else:
        return n*factorial(n-1) #recursive case

#binary search using recursion:
def binarySearch(arr, low, high, x):
    if high >= low :
        mid = (low + high) // 2
        if arr[mid] == x :
            return mid
        elif arr[mid] > x :
            return binarySearch(arr, low, mid - 1, x)
        else:
            return binarySearch(arr, mid + 1, high, x)
    else:
        return -1


#test the functions:
countDown(4)
print(factorial(5))
arr = [1,2,5,8,10,11,12]
print(binarySearch(arr, 0, 6, 10))   #prints: 4
print(binarySearch(arr, 0, 6, 100))  #prints: -1

File: Grokking Algorithms in C++/Chapter 3 .cpp
#include <iostream>
using namespace std;

//function count down from N to 1 using recursion
void countDown(int N)
{
	if (N <= 0)
		return;
	else
	{
		cout << N << " ";
		countDown(N - 1);
	}
}

int factorial(int N)
{
	if (N == 0 || N == 1)
		return 1;
	else
	{
		return N * factorial(N - 1);
	}
}

//binary search using recursion:
int binarySearch(int arr[], int N, int item, int l, int r)
{
	//int low = l, high = r;
	if (r >= l)
	{
		int mid = (l + r) / 2;

		if (arr[mid] == item)
			return mid;
		else if (arr[mid] > item)
			return binarySearch(arr, N, item, l, mid - 1);
		else
			return binarySearch(arr, N, item, mid + 1, r);

	}
	return -1;
}

int main()
{
	countDown(5);
	cout << factorial(5) << endl;
	int myArray[4] = { 1,2,3,4 };
	cout << binarySearch(myArray, 4, 4, 0, 3) << endl;//prints: 3
	cout << binarySearch(myArray, 4, 9, 0, 3) << endl;//prints: -1
	return 0;
}